shell/gpu/gpu_surface_metal.mm

```diff
@@ -112,11 +139,26 @@ GrBackendRenderTarget metal_render_target(bounds.width * scale,   // width
     return nullptr;
   }
 
-  auto submit_callback = [drawable = next_drawable, command_buffer](
+  // External views need to present with transaction. When presenting with
+  // transaction, we have to block, otherwise we risk presenting the drawable
+  // after the CATransaction has completed.
+  // See:
+  // https://developer.apple.com/documentation/quartzcore/cametallayer/1478157-presentswithtransaction
+  // TODO(dnfield): only do this if transactions are actually being used.
+  // https://github.com/flutter/flutter/issues/24133
+  bool wait = delegate_->GetExternalViewEmbedder() != nullptr;
+
+  auto submit_callback = [drawable = next_drawable, command_buffer, wait](
                              const SurfaceFrame& surface_frame, SkCanvas* canvas) -> bool {
     canvas->flush();
-    [command_buffer.get() presentDrawable:drawable.get()];
-    [command_buffer.get() commit];
+    if (!wait) {
+      [command_buffer.get() presentDrawable:drawable.get()];
+      [command_buffer.get() commit];
+    } else {
+      [command_buffer.get() commit];
+      [command_buffer.get() waitUntilScheduled];
```

Comment:
We should do as much work as we possibly can before getting the next drawable - guidelines from Apple say to hold it for as little time as possible. But if we use transactions, we have to block before presenting so that we are sure to present in the transaction. Alternatively, we have to make sure to _not_ commit until we've presented in a callback.